doc/aaa/AAA Improvements/AAA Improvements.md

```diff
@@ -0,0 +1,636 @@
+    
+      And it doesn't end with tracking shell activity. Standard Linux tools that collect metrics about users will not work correctly. For example, **[auditd](http://man7.org/linux/man-pages/man8/auditd.8.html)** is a popular tool that can collect just about anything about users, but it cannot function correctly unless all users have their own Linux credentials.
+  
+  - **Side effect #2 - Users assigned to remote_user_su cannot run sudo.**
+    The sudo program requires users to authenticate themselves before it runs a command. Unfortunately, when we map a RADIUS/TACACS+ user to **remote_user_su**, that user ceases to be the user defined by the RADIUS/TACACS+ server, and it takes on the identity of **remote_user_su**. And since **remote_user_su** does not have login privileges and thus the ability to authenticate, users mapped to **remote_user_su** cannot run commands with sudo.
+  
+- **Authentication for containers is not supported.** Applications running in containers have no visibility of the PAM/NSS configuration that resides on the host. They are entirely unaware of any RADIUS/TACACS+ configuration. The way we make this work today is by establishing an ssh connection from the container to the host. If the connection is successful, then we conclude that the authentication was successful. Using ssh may work for applications that we design ourselves (e.g., REST server), but that won't work with standard Linux applications. For example, when running a bash shell inside a container, one would not be able to be authenticated by sudo or su.
+
+## Proposal for hardening and securing SONiC AAA
+
+This proposal applies to all login methods: Local, RADIUS, and TACACS+  (or even LDAP in the future). It relies on a new NSS module to automatically create user accounts and assign Linux credentials (UID, GID) to users that are missing from `/etc/passwd`. We refer to this as *System-Assigned Credentials* or **sac** for short. 
+
+With **sac**, the Primary GID is no longer used as the **role** and is simply assigned by **sac** as a private GID for that user (refer to the `--user-group` option of the [useradd](http://man7.org/linux/man-pages/man8/useradd.8.html) command). As mentioned before, using the primary GID simply does not work when multiple roles are needed. Instead, the roles are saved to the **REDIS** database when the `pam_setcred()` API is invoked, which is in line with this API's primary purpose (i.e., to set additional credentials). 
```

Comment:
How does the design handles multiple component failures? Ex: HAMD, REDIS, TACACS etc?

Reply:
If the TACACS+ (or RADIUS) servers are down, then obviously any users configured in these servers won't get authenticated and thus won't be able log in. Locally defined users (in /etc/passwd, /etc/group, and /etc/shadow) are not affected.

If `hamd` is down (not sure why/how it would be down) then it's only TACACS+/RADIUS users that don't have local credentials that would not be allowed to log in. Note that once a TACACS+/RADIUS user has previously logged in, their credentials are persisted in the local `/etc/passwd` and they won't need hamd for any future logins. Locally defined users (in /etc/passwd, /etc/group, and /etc/shadow) are not affected.

If REDIS is down (again not sure why/how it would be down), then we won't be able to save the Roles (and/or Privilege Level) and applications won't be able to retrieve the Roles (and/or Privilege Level). Only applications requiring Role-Based Access Control (RBAC) would be affected and they may need to operate with a default, low-level role. 

If `sshd` is down, well obviously nobody would be able to `ssh` into the system.